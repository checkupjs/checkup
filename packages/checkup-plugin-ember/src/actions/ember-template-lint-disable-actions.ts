import { ActionsEvaluator, TaskConfig, sumOccurrences } from '@checkup/core';
import { Result } from 'sarif';

export function evaluateActions(taskResults: Result[], taskConfig: TaskConfig) {
  let actionsEvaluator = new ActionsEvaluator();
  let templateLintDisableUsages = sumOccurrences(taskResults);

  actionsEvaluator.add({
    taskName: 'ember-template-lint-disable',
    name: 'reduce-template-lint-disable-usages',
    summary: 'Reduce number of template-lint-disable usages',
    details: `${templateLintDisableUsages} usages of template-lint-disable`,
    defaultThreshold: 2,
    items: [`Total template-lint-disable usages: ${templateLintDisableUsages}`],
    input: templateLintDisableUsages,
  });

  return actionsEvaluator.evaluate(taskConfig);
}
