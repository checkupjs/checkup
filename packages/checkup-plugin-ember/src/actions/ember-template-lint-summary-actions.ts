import { ActionsEvaluator, TaskConfig, sumOccurrences } from '@checkup/core';
import { Result } from 'sarif';

export function evaluateActions(taskResults: Result[], taskConfig: TaskConfig) {
  let actionsEvaluator = new ActionsEvaluator();

  let errors = taskResults.filter((result: Result) => result.properties?.type === 'error')!;
  let warnings = taskResults.filter((result: Result) => result.properties?.type === 'warning')!;

  let errorCount = sumOccurrences(errors);
  let warningCount = sumOccurrences(warnings);

  actionsEvaluator.add({
    taskName: 'ember-template-lint-summary',
    name: 'reduce-template-lint-errors',
    summary: 'Reduce number of template-lint errors',
    details: `${errorCount} total errors`,
    defaultThreshold: 20,
    items: [`Total template-lint errors: ${errorCount}`],
    input: errorCount,
  });

  actionsEvaluator.add({
    taskName: 'ember-template-lint-summary',
    name: 'reduce-template-lint-warnings',
    summary: 'Reduce number of template-lin warnings',
    details: `${warningCount} total warnings`,
    defaultThreshold: 20,
    items: [`Total template-lint warnings: ${warningCount}`],
    input: warningCount,
  });

  return actionsEvaluator.evaluate(taskConfig);
}
