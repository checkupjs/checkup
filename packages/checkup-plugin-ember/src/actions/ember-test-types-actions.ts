import { ActionsEvaluator, toPercent, TaskConfig, sumOccurrences } from '@checkup/core';
import { Result } from 'sarif';

export function evaluateActions(taskResults: Result[], taskConfig: TaskConfig) {
  let actionsEvaluator = new ActionsEvaluator();

  let totalSkippedTests = sumOccurrences(
    taskResults.filter((taskResult) => taskResult.properties?.method === 'skip')
  );

  let totalTests = sumOccurrences(taskResults);

  actionsEvaluator.add({
    taskName: 'ember-test-types',
    name: 'reduce-skipped-tests',
    summary: 'Reduce number of skipped tests',
    details: `${toPercent(totalSkippedTests, totalTests)} of tests are skipped`,
    defaultThreshold: 0.01,

    items: [`Total skipped tests: ${totalSkippedTests}`],
    input: totalSkippedTests / totalTests,
  });

  return actionsEvaluator.evaluate(taskConfig);
}
