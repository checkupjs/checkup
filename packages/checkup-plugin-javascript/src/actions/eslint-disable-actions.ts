import { TaskResult, ActionsEvaluator } from '@checkup/core';

export function evaluateActions(taskResult: TaskResult) {
  let actionsEvaluator = new ActionsEvaluator();
  let summaryResult = taskResult.data[0];
  let eslintDisableUsages = summaryResult.count;

  actionsEvaluator.add({
    name: 'reduce-eslint-disable-usages',
    summary: 'Reduce number of eslint-disable usages',
    details: `${eslintDisableUsages} usages of template-lint-disable`,
    defaultThreshold: 2,
    items: [`Total eslint-disable usages: ${eslintDisableUsages}`],
    input: eslintDisableUsages,
  });

  return actionsEvaluator.evaluate(taskResult.config);
}
