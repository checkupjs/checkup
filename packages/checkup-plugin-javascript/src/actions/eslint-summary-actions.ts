import { TaskResult, ActionsEvaluator, MultiValueResult } from '@checkup/core';

export function evaluateActions(taskResult: TaskResult) {
  let actionsEvaluator = new ActionsEvaluator();

  let errors = taskResult.data.find((result: MultiValueResult) => result.key === 'eslint-errors')!;
  let warnings = taskResult.data.find(
    (result: MultiValueResult) => result.key === 'eslint-warnings'
  )!;

  let errorCount = errors.dataSummary.total;
  let warningCount = warnings.dataSummary.total;

  actionsEvaluator.add({
    name: 'reduce-eslint-errors',
    summary: 'Reduce number of eslint errors',
    details: `${errorCount} total errors`,
    defaultThreshold: 20,
    items: [`Total eslint errors: ${errorCount}`],
    input: errorCount,
  });

  actionsEvaluator.add({
    name: 'reduce-eslint-warnings',
    summary: 'Reduce number of eslint warnings',
    details: `${warningCount} total warnings`,
    defaultThreshold: 20,
    items: [`Total eslint warnings: ${warningCount}`],
    input: warningCount,
  });

  return actionsEvaluator.evaluate(taskResult.config);
}
