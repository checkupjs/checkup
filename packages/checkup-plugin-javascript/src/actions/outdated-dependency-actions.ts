import { ActionsEvaluator, toPercent, TaskConfig, sumOccurrences } from '@checkup/core';
import { Result } from 'sarif';

export function evaluateActions(taskResults: Result[], taskConfig: TaskConfig) {
  let totalDependencies = sumOccurrences(taskResults);

  let outdatedDependencies = {
    major: sumOccurrences(taskResults.filter((taskResult) => taskResult.message.text === 'major')),
    minor: sumOccurrences(taskResults.filter((taskResult) => taskResult.message.text === 'minor')),
  };

  let actionsEvaluator = new ActionsEvaluator();

  actionsEvaluator.add({
    name: 'reduce-outdated-major-dependencies',
    summary: 'Update outdated major versions',
    details: `${outdatedDependencies.major} major versions outdated`,
    defaultThreshold: 0.05,
    items: [],
    input: outdatedDependencies.major / totalDependencies,
  });
  actionsEvaluator.add({
    name: 'reduce-outdated-minor-dependencies',
    summary: 'Update outdated minor versions',
    details: `${outdatedDependencies.minor} minor versions outdated`,
    defaultThreshold: 0.05,
    items: [],
    input: outdatedDependencies.minor / totalDependencies,
  });
  actionsEvaluator.add({
    name: 'reduce-outdated-dependencies',
    summary: 'Update outdated versions',
    details: `${toPercent(
      (outdatedDependencies.major + outdatedDependencies.minor) / totalDependencies
    )} of versions outdated`,
    defaultThreshold: 0.2,
    items: [],
    input: (outdatedDependencies.major + outdatedDependencies.minor) / totalDependencies,
  });

  return actionsEvaluator.evaluate(taskConfig);
}
