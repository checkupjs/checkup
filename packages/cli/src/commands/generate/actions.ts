import * as yargs from 'yargs';
import Generator from '../../api/generator';
import { consoleWriter } from '../../checkup';

export const generateActionsCommand = {
  command: 'actions <name> [options]',
  describe: 'Generates checkup actions within a project',
  builder: (yargs: any) => {
    return yargs
      .positional('name', {
        description: 'Name of the actions (foo-task-actions)',
        default: '',
      })
      .options({
        defaults: {
          alias: 'd',
          description: 'Use defaults for every setting',
          boolean: true,
        },
        path: {
          alias: 'p',
          default: '.',
          description: 'The path referring to the directory that the generator will run in',
        },
      });
  },
  handler: async (argv: yargs.Arguments) => {
    try {
      let generator = new Generator({
        path: argv.path as string,
        generator: 'actions',
        name: argv.name as string,
        defaults: argv.defaults as boolean,
      });

      await generator.run();
    } catch (error) {
      consoleWriter.error(error);
    }
  },
};
