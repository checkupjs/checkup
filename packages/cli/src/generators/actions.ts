import * as _ from 'lodash';
import * as chalk from 'chalk';
import * as t from '@babel/types';
import * as recast from 'recast';

import traverse from '@babel/traverse';
import { Answers } from 'inquirer';
import BaseGenerator, { Works } from './base-generator';
import { Options } from '../commands/generate';
import { join } from 'path';
import { AstTransformer, CheckupError } from '@checkup/core';

interface ActionOptions extends Options {
  taskName: string;
  pascalCaseName: string;
  typescript: boolean;
}

export default class ActionsGenerator extends BaseGenerator {
  works: Works = Works.InsidePlugin;
  answers!: Answers;

  constructor(args: any, public options: ActionOptions) {
    super(args, options);
  }

  initializing() {
    if (!this.canRunGenerator) {
      throw new CheckupError(
        `Can only generate actions from inside a Checkup plugin directory`,
        `Run ${chalk.bold.white(
          'checkup generate actions'
        )} from the root of a Checkup plugin or use the ${chalk.bold.white(
          '--path'
        )} option to specify the path to a Checkup plugin`
      );
    }
  }

  async prompting() {
    this.headline(`${this.options.name}-actions`);

    const defaults = {
      typescript: true,
    };

    if (this.options.defaults) {
      this.answers = defaults;
    } else {
      this.answers = await this.prompt([
        {
          type: 'confirm',
          name: 'typescript',
          message: 'TypeScript',
          default: () => true,
        },
        {
          type: 'input',
          name: 'taskName',
          message: `Enter the task name that these actions will be associated with (the string value in the taskName property of the task class).`,
        },
      ]);
    }

    this.options.taskName = this.answers.taskName;
    this.options.pascalCaseName = _.upperFirst(_.camelCase(this.options.name));
    this.options.typescript = this.answers.typescript;
  }

  writing() {
    this.sourceRoot(join(__dirname, '../../templates/src/actions'));

    if (
      !this.fs.exists(
        this.destinationPath(`${this._dir}/registrations/register-actions.${this._ext}`)
      )
    ) {
      this.fs.copy(
        this.templatePath(`src/registrations/register-actions.${this._ext}.ejs`),
        this.destinationPath(`${this._dir}/registrations/register-actions.${this._ext}`)
      );
    }

    this.fs.copyTpl(
      this.templatePath(`src/actions/actions.${this._ext}.ejs`),
      this.destinationPath(`${this._dir}/actions/${this.options.name}-actions.${this._ext}`),
      this.options
    );

    this._transformHooks();
  }

  private _transformHooks() {
    let hooksDestinationPath = this.destinationPath(
      `${this._dir}/registrations/register-actions.${this._ext}`
    );

    let registerActionsSource = this.fs.read(hooksDestinationPath);
    let registerActionStatement = t.expressionStatement(
      t.callExpression(t.identifier('registerActions'), [
        t.stringLiteral(this.options.taskName),
        t.identifier(`evaluate${this.options.pascalCaseName}Actions`),
      ])
    );

    let newActionImportSpecifier = t.importSpecifier(
      t.identifier(`evaluate${this.options.pascalCaseName}Actions`),
      t.identifier('evaluateActions')
    );
    let tasksImportDeclaration: t.ImportDeclaration = t.importDeclaration(
      [newActionImportSpecifier],
      t.stringLiteral(`../actions/${this.options.name}-actions`)
    );

    let code = new AstTransformer(registerActionsSource, recast.parse, traverse, {
      parser: require('recast/parsers/typescript'),
    })
      .traverse({
        Program(path) {
          path.node.body.splice(1, 0, tasksImportDeclaration);
        },
        BlockStatement(path) {
          path.node.body.push(registerActionStatement);
        },
      })
      .generate();

    this.fs.write(hooksDestinationPath, code);
  }
}
