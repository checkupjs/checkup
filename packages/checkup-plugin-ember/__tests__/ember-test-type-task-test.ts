import { EmberProject, getTaskContext } from '@checkup/test-helpers';
import { getPluginName } from '@checkup/core';

import EmberTestTypesTask from '../src/tasks/ember-test-types-task';
import { evaluateActions } from '../src/actions/ember-test-types-actions';

const TESTS = {
  application: {
    tests: {
      'my-component-test.js': `
      module('test', function(hooks) {
        setupApplicationTest(hooks);

        test('testing foo');
        skip('testing shmoo')
        skip('testing shmoo')
      });
    `,
    },
    someFolder: {
      'my-component-test.js': `
      module('test', function(hooks) {
        setupApplicationTest(hooks);

        test('testing foo');
        skip('testing shmoo')
        skip('testing shmoo')
      });
    `,
    },
  },
  rendering: {
    tests: {
      'my-component-test.js': `
      module('test', function(hooks) {
        setupRenderingTest(hooks);

        test('testing foo');
        skip('testing shmoo')
        skip('testing shmoo')
      });
    `,
    },
  },
  unit: {
    tests: {
      'my-component-test.js': `
      module('test', function(hooks) {
        setupTest(hooks);

        test('testing foo');
        skip('testing shmoo')
        skip('testing shmoo')
      });
    `,
    },
  },
};

describe('ember-test-types-task', () => {
  let project: EmberProject;
  let pluginName = getPluginName(__dirname);

  beforeEach(function () {
    project = new EmberProject('checkup-app', '0.0.0');
  });

  afterEach(function () {
    project.dispose();
  });

  it('only renders todos/only test types if they a value > 0', async () => {
    project.files = Object.assign(project.files, {
      'index.js': 'index js file',
      addon: {
        rendering: {
          tests: {
            'my-component-test.js': `
              module('test', function(hooks) {
                setupRenderingTest(hooks);

                test('testing foo');
                only('testing shmoo')
                todo('testing shmoo')
              });
            `,
          },
        },
      },
    });

    project.writeSync();

    const result = await new EmberTestTypesTask(
      pluginName,
      getTaskContext({ options: { cwd: project.baseDir }, paths: project.filePaths })
    ).run();

    expect(result).toMatchSnapshot();
  });

  it('returns all the test types found in the app and outputs to json', async () => {
    project.files = Object.assign(project.files, {
      'index.js': 'index js file',
      addon: TESTS,
    });

    project.writeSync();

    const result = await new EmberTestTypesTask(
      pluginName,
      getTaskContext({ options: { cwd: project.baseDir }, paths: project.filePaths })
    ).run();

    expect(result).toMatchSnapshot();
  });

  it('returns action item if more than 1% of your tests are skipped and if your ratio of application tests is not matching threshold', async () => {
    project.files = Object.assign(project.files, {
      'index.js': 'index js file',
      addon: TESTS,
    });

    project.writeSync();

    const task = new EmberTestTypesTask(
      pluginName,
      getTaskContext({
        options: { cwd: project.baseDir },
        paths: project.filePaths,
        config: {
          tasks: {
            'ember/ember-test-types': [
              'on',
              {
                actions: {
                  'ratio-application-tests': [
                    'on',
                    {
                      threshold: 3,
                    },
                  ],
                },
              },
            ],
          },
        },
      })
    );
    const result = await task.run();

    let actions = evaluateActions(result, task.config);

    expect(actions).toHaveLength(1);
    expect(actions).toMatchInlineSnapshot(`
      Array [
        Object {
          "defaultThreshold": 0.01,
          "details": "67% of tests are skipped",
          "input": 0.6666666666666666,
          "items": Array [
            "Total skipped tests: 8",
          ],
          "name": "reduce-skipped-tests",
          "summary": "Reduce number of skipped tests",
        },
      ]
    `);
  });
});
