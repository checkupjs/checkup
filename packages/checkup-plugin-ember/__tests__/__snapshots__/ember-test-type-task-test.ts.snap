// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ember-test-types-task only renders todos/only test types if they a value > 0 1`] = `
"Ember Test Types

Type           test    only    todo    
unit           0       0       0       
rendering      1       1       1       
application    0       0       0       

Test Type Breakdown

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 tests
■ unit (0)  ■ rendering (3)  ■ application (0)  

"
`;

exports[`ember-test-types-task returns all the test types (including nested) found in the app and outputs to the console 1`] = `
"Ember Test Types

Type           test    skip    
unit           2       4 (67%) 
rendering      2       4 (67%) 
application    4       8 (67%) 

Test Type Breakdown

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 24 tests
■ unit (6)  ■ rendering (6)  ■ application (12)  

"
`;

exports[`ember-test-types-task returns all the test types found in the app and outputs to json 1`] = `
Object {
  "info": Object {
    "friendlyTaskName": "Ember Test Types",
    "taskClassification": Object {
      "category": "testing",
      "group": "ember",
    },
    "taskName": "ember-test-types",
  },
  "result": Object {
    "types": Array [
      Object {
        "only": 0,
        "skip": 2,
        "test": 1,
        "todo": 0,
        "total": 3,
        "type": "unit",
      },
      Object {
        "only": 0,
        "skip": 2,
        "test": 1,
        "todo": 0,
        "total": 3,
        "type": "rendering",
      },
      Object {
        "only": 0,
        "skip": 4,
        "test": 2,
        "todo": 0,
        "total": 6,
        "type": "application",
      },
    ],
  },
}
`;

exports[`ember-test-types-task returns all the test types found in the app and outputs to the console 1`] = `
"Ember Test Types

Type           test    skip    
unit           1       2 (67%) 
rendering      1       2 (67%) 
application    2       4 (67%) 

Test Type Breakdown

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 12 tests
■ unit (3)  ■ rendering (3)  ■ application (6)  

"
`;
